<a href="https://colab.research.google.com/github/anabiarochar/Mestrado_AlgoritmosProgramacao/blob/main/Projeto1/Inser%C3%A7%C3%A3o.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [34]:
import requests
import time

# 1. Obter os dados do GitHub
url = "https://raw.githubusercontent.com/anabiarochar/Mestrado_AlgoritmosProgramacao/refs/heads/main/Projeto1/Vetores/vetor_100000_decrescente.txt?token=GHSAT0AAAAAADXG3NI5MPMPFAIQCL7U5CGI2NNGDRQ"

try:
    response = requests.get(url)
    content = response.text

    # Lendo o arquivo separado por vírgulas
    vetor = [int(x.strip()) for x in content.split(',') if x.strip()]

    print(f"Sucesso! Vetor carregado com {len(vetor)} elementos.")

except Exception as e:
    print(f"Erro ao carregar ou processar o arquivo: {e}")
    vetor = []

# 2. Implementação do Método de Inserção (Insertion Sort)
def insertion_sort(lista_original):
    # Criamos uma cópia para preservar o vetor original
    arr = lista_original.copy()
    n = len(arr)

    # Começamos do segundo elemento (índice 1)
    for i in range(1, n):
        chave = arr[i]
        j = i - 1

        # Move os elementos de arr[0..i-1] que são maiores que a chave
        # para uma posição à frente de sua posição atual
        while j >= 0 and chave < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1

        # Insere a chave na posição correta
        arr[j + 1] = chave

    return arr

# 3. Execução e Medição de Tempo
if vetor:
    print("\nIniciando ordenação por Insertion Sort...")

    # Início do cronômetro de alta precisão
    inicio = time.perf_counter()

    vetor_ordenado = insertion_sort(vetor)

    # Fim do cronômetro
    fim = time.perf_counter()

    tempo_total = fim - inicio

    print("--- Resultado ---")
    print(f"Tempo de execução: {tempo_total:.6f} segundos")
    print(f"Primeiros 10 elementos ordenados: {vetor_ordenado[:10]}")
    print(f"Últimos 10 elementos ordenados: {vetor_ordenado[-10:]}")
else:
    print("Não foi possível processar o vetor.")

Sucesso! Vetor carregado com 100000 elementos.

Iniciando ordenação por Insertion Sort...
--- Resultado ---
Tempo de execução: 348.336984 segundos
Primeiros 10 elementos ordenados: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Últimos 10 elementos ordenados: [99991, 99992, 99993, 99994, 99995, 99996, 99997, 99998, 99999, 100000]


###Aleatório 1000:
Tentativa 1 - Tempo de execução: 0.033426 segundos  
Tentativa 2 - Tempo de execução: 0.015758 segundos  
Tentativa 3 - Tempo de execução: 0.016101 segundos  
Tentativa 4 - Tempo de execução: 0.016582 segundos  

###Aleatório 10000:
Tentativa 1 - Tempo de execução: 1.780963 segundos  
Tentativa 2 - Tempo de execução: 1.683920 segundos  
Tentativa 3 - Tempo de execução: 1.958337 segundos  
Tentativa 4 - Tempo de execução: 2.279688 segundos  

###Aleatório 100000:
Tentativa 1 - Tempo de execução: 400.612380 segundos   
Tentativa 2 - Tempo de execução: 455.484334 segundos  
Tentativa 3 - Tempo de execução:  
Tentativa 4 - Tempo de execução:  

###Crescente 1000:
Tentativa 1 - Tempo de execução: 0.006639 segundos    
Tentativa 2 - Tempo de execução: 0.000185 segundos  
Tentativa 3 - Tempo de execução: 0.000141 segundos  
Tentativa 4 - Tempo de execução: 0.000107 segundos

###Crescente 10000:
Tentativa 1 - Tempo de execução: 0.000928 segundos  
Tentativa 2 - Tempo de execução: 0.001567 segundos  
Tentativa 3 - Tempo de execução: 0.000901 segundos  
Tentativa 4 - Tempo de execução: 0.000945 segundos  

###Crescente 100000:
Tentativa 1 - Tempo de execução: 0.010484 segundos  
Tentativa 2 - Tempo de execução: 0.008961 segundos  
Tentativa 3 - Tempo de execução: 0.009683 segundos  
Tentativa 4 - Tempo de execução: 0.009491 segundos  

###Decrescente 1000:
Tentativa 1 - Tempo de execução: 0.029543 segundos  
Tentativa 2 - Tempo de execução: 0.047466 segundos  
Tentativa 3 - Tempo de execução: 0.047212 segundos  
Tentativa 4 - Tempo de execução: 0.043740 segundos   

###Decrescente 10000:
Tentativa 1 - Tempo de execução: 3.203136 segundos   
Tentativa 2 - Tempo de execução: 3.738098 segundos   
Tentativa 3 - Tempo de execução: 3.283708 segundos   
Tentativa 4 - Tempo de execução: 3.225976 segundos   

###Decrescente 100000:
Tentativa 1 - Tempo de execução: 349.920139 segundos    
Tentativa 2 - Tempo de execução: 348.336984 segundos   
Tentativa 3 - Tempo de execução:  
Tentativa 4 - Tempo de execução:  